File: src/personal_os/source_conflicts/contracts.py
"""Framework-neutral source-conflict domain contracts (Child 8 spec 4).

Closed vocabularies and immutable value objects for the
server-authoritative conflict aggregate: the four conflict kinds, the four
aggregate states of the capture/resolve state machine, the three explicit
resolution choices, the strict event idempotency key, the content-versus-
delete candidate shape of the spec 4.1 table, and the safe aggregate read
model capture, read and replay return. The module imports no FastAPI,
SQLAlchemy, R2 or request type; device and workspace identity arrive only
through the issuing domain's credential-derived context, never from capture
data. Locators, digests, tokens, candidate bytes and merged drafts are
private values: they never enter diagnostics, error details or metric
labels, and the locator snapshot exists only to cross the store boundary
into canonical state.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from datetime import datetime
from enum import StrEnum
from typing import Final
from uuid import UUID

from personal_os.sources.actors import reject_nil_uuid
from personal_os.sources.commands import normalize_utc_timestamp

#: Canonical idempotency-key grammar: exactly the canonical lowercase
#: hyphenated UUID text form the plugin mints with ``crypto.randomUUID``.
#: Capture and resolution each carry one fresh key bound to one event
#: identity; an exact replay of that identity returns the stored outcome.
_IDEMPOTENCY_KEY_PATTERN: Final[re.Pattern[str]] = re.compile(
    r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$"
)


class ConflictKind(StrEnum):
    """Closed vocabulary of conflict kinds (spec 4.1 table).

    ``STALE_CONTENT`` and ``EDIT_REMOTE_DELETE`` retain verified content
    bytes; ``DELETE_REMOTE_EDIT`` carries only a deletion intent; a
    ``LOCATOR_COLLISION`` carries a locator snapshot and retains content
    bytes only when the local bytes changed.
    """

    STALE_CONTENT = "stale_content"
    EDIT_REMOTE_DELETE = "edit_remote_delete"
    DELETE_REMOTE_EDIT = "delete_remote_edit"
    LOCATOR_COLLISION = "locator_collision"


class ConflictStatus(StrEnum):
    """Closed aggregate states of the conflict state machine (spec 4.3)."""

    OPEN = "open"
    RESOLVING = "resolving"
    RESOLVED = "resolved"
    SUPERSEDED = "superseded"


#: The two terminal states: they accept no further mutation and exact replay
#: returns them unchanged, never a duplicate winner.
TERMINAL_CONFLICT_STATUSES: Final[frozenset[ConflictStatus]] = frozenset(
    {ConflictStatus.RESOLVED, ConflictStatus.SUPERSEDED}
)


class ConflictResolutionKind(StrEnum):
    """The only explicit user choices that close a conflict (spec 3.8)."""

    KEEP_REMOTE = "keep_remote"
    KEEP_LOCAL = "keep_local"
    SAVE_MERGED = "save_merged"


#: The resolution choices that publish exactly one immutable source version
#: against the reviewed remote version; ``keep_remote`` publishes none.
VERSION_PUBLISHING_RESOLUTIONS: Final[frozenset[ConflictResolutionKind]] = frozenset(
    {ConflictResolutionKind.KEEP_LOCAL, ConflictResolutionKind.SAVE_MERGED}
)


class ConflictResolutionOutcome(StrEnum):
    """Closed terminal outcomes of one resolution attempt.

    ``RESOLVED`` commits the winner; ``STALE_SUCCESSOR`` records the attempt
    as stale against the reviewed remote, supersedes the conflict and opens
    a successor bound to the newer observed remote. Both are frozen and
    returned unchanged by an exact replay of the resolution event identity.
    """

    RESOLVED = "resolved"
    STALE_SUCCESSOR = "stale_successor"


class ConflictCandidateKind(StrEnum):
    """Whether a conflict retains verified content bytes or a deletion intent."""

    CONTENT = "content"
    DELETE = "delete"


class ConflictEvidenceRole(StrEnum):
    """The three immutable evidence roles the verified-read boundary streams.

    ``CANDIDATE`` exists only while the conflict retains a content candidate;
    a delete conflict has no candidate bytes and the reader fails closed.
    """

    BASE = "base"
    REMOTE = "remote"
    CANDIDATE = "candidate"


@dataclass(frozen=True, slots=True)
class ConflictIdempotencyKey:
    """Stable event idempotency key in canonical UUID text form.

    The grammar accepts only the canonical lowercase hyphenated 8-4-4-4-12
    form; uppercase, braces, ``urn:uuid:`` prefixes, whitespace, wrong
    lengths, the nil UUID and any non-printable form are rejected without
    normalization. The key is opaque, event-scoped and never logged.
    """

    value: str

    def __repr__(self) -> str:
        return f"{type(self).__name__}(value=<redacted>)"

    def __post_init__(self) -> None:
        if _IDEMPOTENCY_KEY_PATTERN.fullmatch(self.value) is None:
            raise ValueError("idempotency key must be a canonical lowercase hyphenated UUID")
        if self.value == "00000000-0000-0000-0000-000000000000":
            raise ValueError("idempotency key must be a non-nil UUID")


@dataclass(frozen=True, slots=True)
class ConflictCandidate:
    """One conflict's retained local-side evidence shape (spec 4.1 table).

    A content candidate is the immutable verified object the existing
    single-part or multipart flow already proved by digest, size and media
    type — never raw bytes and never a source version. A delete candidate
    is a no-byte deletion intent; it must not carry a content object, so a
    conflict can never be partly a deletion and partly a content candidate.
    """

    candidate_kind: ConflictCandidateKind
    verified_candidate_object_id: UUID | None

    def __post_init__(self) -> None:
        if self.candidate_kind is ConflictCandidateKind.CONTENT:
            if self.verified_candidate_object_id is None:
                raise ValueError("a content candidate requires a verified_candidate_object_id")
            reject_nil_uuid("verified_candidate_object_id", self.verified_candidate_object_id)
            return
        if self.verified_candidate_object_id is not None:
            raise ValueError("a delete candidate must not carry a verified_candidate_object_id")

    @classmethod
    def content(cls, verified_candidate_object_id: UUID | None) -> ConflictCandidate:
        """Build the content variant; the verified object reference is required."""
        return cls(
            candidate_kind=ConflictCandidateKind.CONTENT,
            verified_candidate_object_id=verified_candidate_object_id,
        )

    @classmethod
    def delete(cls, verified_candidate_object_id: UUID | None = None) -> ConflictCandidate:
        """Build the no-byte deletion variant; a content object is refused."""
        return cls(
            candidate_kind=ConflictCandidateKind.DELETE,
            verified_candidate_object_id=verified_candidate_object_id,
        )


def validate_candidate_for_kind(conflict_kind: ConflictKind, candidate: ConflictCandidate) -> None:
    """Enforce the candidate requirement of the spec 4.1 table.

    ``stale_content`` and ``edit_remote_delete`` conflicts require a content
    candidate; a ``delete_remote_edit`` conflict requires a delete candidate;
    a ``locator_collision`` accepts either shape because its content object
    is required only when the local bytes changed.
    """

    if (
        conflict_kind in {ConflictKind.STALE_CONTENT, ConflictKind.EDIT_REMOTE_DELETE}
        and candidate.candidate_kind is not ConflictCandidateKind.CONTENT
    ):
        raise ValueError(f"a {conflict_kind.value} conflict requires a content candidate")
    if (
        conflict_kind is ConflictKind.DELETE_REMOTE_EDIT
        and candidate.candidate_kind is not ConflictCandidateKind.DELETE
    ):
        raise ValueError(f"a {conflict_kind.value} conflict requires a delete candidate")


@dataclass(frozen=True, slots=True)
class SourceConflict:
    """The frozen aggregate read model capture, read and replay return.

    Carries only opaque identifiers, closed labels and normalized UTC
    timestamps: the locator snapshot lives in canonical state and the raw
    locator, candidate bytes, digests and object keys never cross into this
    value. Status shapes are exact: an open conflict carries no resolution
    evidence; a resolving conflict carries the attempt identity but no
    outcome; a resolved conflict binds the winner and — only under
    ``keep_local``/``save_merged`` — exactly one resulting version; a
    superseded conflict binds the stale attempt and its successor and never
    a resulting version. Evidence is immutable: no field describes later
    remote state.
    """

    conflict_id: UUID
    workspace_id: UUID
    source_id: UUID | None
    conflict_kind: ConflictKind
    status: ConflictStatus
    originating_event_id: UUID
    originating_device_id: UUID
    base_version_id: UUID | None
    observed_remote_version_id: UUID | None
    candidate: ConflictCandidate
    captured_at: datetime
    resolution_kind: ConflictResolutionKind | None
    resolution_event_id: UUID | None
    resulting_version_id: UUID | None
    successor_conflict_id: UUID | None
    closed_at: datetime | None

    def __post_init__(self) -> None:
        if not isinstance(self.conflict_kind, ConflictKind):
            raise ValueError("conflict_kind must be a closed ConflictKind")
        if not isinstance(self.status, ConflictStatus):
            raise ValueError("status must be a closed ConflictStatus")
        if not isinstance(self.candidate, ConflictCandidate):
            raise ValueError("candidate must be a ConflictCandidate value")
        reject_nil_uuid("conflict_id", self.conflict_id)
        reject_nil_uuid("workspace_id", self.workspace_id)
        reject_nil_uuid("originating_event_id", self.originating_event_id)
        reject_nil_uuid("originating_device_id", self.originating_device_id)
        if self.base_version_id is not None:
            reject_nil_uuid("base_version_id", self.base_version_id)
        if self.observed_remote_version_id is not None:
            reject_nil_uuid("observed_remote_version_id", self.observed_remote_version_id)
        if self.source_id is not None:
            reject_nil_uuid("source_id", self.source_id)
        elif self.conflict_kind is not ConflictKind.LOCATOR_COLLISION:
            raise ValueError("source_id may be null only for a locator_collision conflict")
        if self.resolution_event_id is not None:
            reject_nil_uuid("resolution_event_id", self.resolution_event_id)
        if self.resulting_version_id is not None:
            reject_nil_uuid("resulting_version_id", self.resulting_version_id)
        if self.successor_conflict_id is not None:
            reject_nil_uuid("successor_conflict_id", self.successor_conflict_id)
        validate_candidate_for_kind(self.conflict_kind, self.candidate)
        object.__setattr__(
            self, "captured_at", normalize_utc_timestamp("captured_at", self.captured_at)
        )
        object.__setattr__(self, "closed_at", normalize_utc_timestamp("closed_at", self.closed_at))
        self._validate_status_shape()

    def _validate_status_shape(self) -> None:
        if self.status is ConflictStatus.OPEN:
            if (
                self.resolution_kind is not None
                or self.resolution_event_id is not None
                or self.resulting_version_id is not None
                or self.successor_conflict_id is not None
                or self.closed_at is not None
            ):
                raise ValueError("an open conflict carries no resolution evidence")
            return
        if self.status is ConflictStatus.RESOLVING:
            if self.resolution_kind is None or self.resolution_event_id is None:
                raise ValueError(
                    "a resolving conflict requires resolution_kind and resolution_event_id"
                )
            if (
                self.resulting_version_id is not None
                or self.successor_conflict_id is not None
                or self.closed_at is not None
            ):
                raise ValueError(
                    "a resolving conflict carries no resulting version, successor or closed_at"
                )
            return
        if self.status is ConflictStatus.RESOLVED:
            if (
                self.resolution_kind is None
                or self.resolution_event_id is None
                or self.closed_at is None
            ):
                raise ValueError(
                    "a resolved conflict requires resolution_kind, resolution_event_id "
                    "and closed_at"
                )
            if self.successor_conflict_id is not None:
                raise ValueError("a resolved conflict carries no successor")
            if self.resolution_kind in VERSION_PUBLISHING_RESOLUTIONS:
                if self.resulting_version_id is None:
                    raise ValueError(
                        f"a {self.resolution_kind.value} resolution requires a resulting_version_id"
                    )
            elif self.resulting_version_id is not None:
                raise ValueError("a keep_remote resolution creates no source version")
            return
        if (
            self.resolution_event_id is None
            or self.successor_conflict_id is None
            or self.closed_at is None
        ):
            raise ValueError(
                "a superseded conflict requires resolution_event_id, successor_conflict_id "
                "and closed_at"
            )
        if self.resulting_version_id is not None:
            raise ValueError("a superseded conflict carries no resulting_version_id")
        if self.successor_conflict_id == self.conflict_id:
            raise ValueError("successor_conflict_id must differ from conflict_id")


__all__ = [
    "TERMINAL_CONFLICT_STATUSES",
    "VERSION_PUBLISHING_RESOLUTIONS",
    "ConflictCandidate",
    "ConflictCandidateKind",
    "ConflictEvidenceRole",
    "ConflictIdempotencyKey",
    "ConflictKind",
    "ConflictResolutionKind",
    "ConflictResolutionOutcome",
    "ConflictStatus",
    "SourceConflict",
    "validate_candidate_for_kind",
]

File: apps/obsidian-plugin/src/conflicts/contracts.ts
/**
 * The strict decoded Conflict Inbox contract surface (Child 8 spec 6,
 * Task 7).
 *
 * The generated workspace client is intentionally NOT bundled into
 * Obsidian, so this module owns the plugin-side mirror of exactly the
 * four Task 6 conflict wire shapes: every closed vocabulary is
 * compile-time bound to the generated server registry through
 * `satisfies`, and every wire record decodes onto a frozen camelCase
 * plugin shape that carries ONLY opaque identifiers, closed labels and
 * normalized timestamps — never a locator, object key, digest, provider
 * detail or any raw response text. A decoder rejects an unknown enum
 * value, an ill-typed member or a foreign payload shape with the single
 * closed {@link ConflictContractError} reason; raw response detail never
 * reaches the thrown failure.
 *
 * The same module owns the closed vocabularies of the durable no-byte
 * repair facts of journal schema v9 (`conflict_local_repairs`): the
 * target-action and safe-reason tokens the journal DDL renders its CHECK
 * constraints from (single source of truth with
 * `../journal/sqlite-database.ts`), and the {@link PendingLocalApply}
 * record whose members make byte or path storage unrepresentable — no
 * `Uint8Array`, blob, path or locator member exists on any input type
 * here.
 *
 * This module carries NO runtime dependency on the journal engine, Node,
 * Electron or the Obsidian adapter — only a type-level import of the
 * generated contract package — so the whole `src/conflicts` tree stays
 * loadable on mobile.
 */

import type { components } from "@workspace/api-client";

// --- the closed wire vocabularies (Task 6 registry binding) ----------------------------------------

/**
 * The closed conflict kind vocabulary: every member is a registered
 * `ConflictKind` of the generated server error registry, so an
 * unregistered snake_case kind fails `tsc --noEmit` here, not at a
 * diagnostics surface.
 */
export const CONFLICT_KINDS = [
  "stale_content",
  "edit_remote_delete",
  "delete_remote_edit",
  "locator_collision",
] as const satisfies readonly components["schemas"]["ConflictKind"][];

/** One conflict's closed kind label. */
export type ConflictKind = (typeof CONFLICT_KINDS)[number];

/** The closed conflict lifecycle status vocabulary. */
export const CONFLICT_STATUSES = [
  "open",
  "resolving",
  "resolved",
  "superseded",
] as const satisfies readonly components["schemas"]["ConflictStatus"][];

/** One conflict's closed lifecycle status. */
export type ConflictStatus = (typeof CONFLICT_STATUSES)[number];

/** The closed candidate shape vocabulary of one conflict. */
export const CONFLICT_CANDIDATE_KINDS = [
  "content",
  "delete",
] as const satisfies readonly components["schemas"]["ConflictCandidateKind"][];

/** One conflict candidate's closed shape (retained bytes or byteless intent). */
export type ConflictCandidateKind = (typeof CONFLICT_CANDIDATE_KINDS)[number];

/** The closed explicit-resolution choice vocabulary. */
export const CONFLICT_RESOLUTION_KINDS = [
  "keep_remote",
  "keep_local",
  "save_merged",
] as const satisfies readonly components["schemas"]["ConflictResolutionKind"][];

/** One explicit resolution's closed choice. */
export type ConflictResolutionKind = (typeof CONFLICT_RESOLUTION_KINDS)[number];

/** The closed resolution outcome vocabulary (`stale_successor` is a typed success). */
export const CONFLICT_RESOLUTION_OUTCOMES = [
  "resolved",
  "stale_successor",
] as const satisfies readonly components["schemas"]["ConflictResolutionOutcome"][];

/** One resolution attempt's closed outcome. */
export type ConflictResolutionOutcome = (typeof CONFLICT_RESOLUTION_OUTCOMES)[number];

/** The closed immutable evidence role vocabulary. */
export const CONFLICT_EVIDENCE_ROLES = [
  "base",
  "remote",
  "candidate",
] as const satisfies readonly components["schemas"]["ConflictEvidenceRole"][];

/** One immutable evidence role of the verified evidence download. */
export type ConflictEvidenceRole = (typeof CONFLICT_EVIDENCE_ROLES)[number];

/** Whether one value is a member of the closed evidence role vocabulary. */
export function isConflictEvidenceRole(value: unknown): value is ConflictEvidenceRole {
  return (
    typeof value === "string" && (CONFLICT_EVIDENCE_ROLES as readonly string[]).includes(value)
  );
}

// --- the decoded wire records -----------------------------------------------------------------------

/**
 * One conflict's safe metadata: the decoded mirror of the server's
 * `SourceConflictData`. Every member is an opaque identifier, a closed
 * label or a normalized UTC timestamp — the credential-derived workspace
 * is the caller's own and no locator, object key, digest or provider
 * detail ever crosses.
 */
export interface ConflictSummary {
  readonly conflictId: string;
  readonly sourceId: string | null;
  readonly conflictKind: ConflictKind;
  readonly status: ConflictStatus;
  readonly originatingEventId: string;
  readonly originatingDeviceId: string;
  readonly baseVersionId: string | null;
  readonly observedRemoteVersionId: string | null;
  readonly candidateKind: ConflictCandidateKind;
  readonly verifiedCandidateObjectId: string | null;
  readonly capturedAt: string;
  readonly resolutionKind: ConflictResolutionKind | null;
  readonly resolutionEventId: string | null;
  readonly resultingVersionId: string | null;
  readonly successorConflictId: string | null;
  readonly closedAt: string | null;
}

/**
 * One conflict's detail: the safe metadata plus exactly the offered
 * choices the server's kind/media-type matrix admits, so the Inbox can
 * never offer an unappliable choice.
 */
export interface ConflictDetail extends ConflictSummary {
  readonly choices: readonly ConflictResolutionKind[];
}

/**
 * One bounded page of the workspace's open conflicts with its stable
 * continuation cursor.
 */
export interface ConflictPage {
  readonly conflicts: readonly ConflictSummary[];
  readonly hasMore: boolean;
  readonly nextExclusiveStartConflictId: string | null;
}

/**
 * The frozen outcome of one explicit resolution attempt: `resolved`
 * commits the winner (with a resulting version only under a publishing
 * choice) and `stale_successor` binds the open successor created against
 * the newer observed remote; a same-identity replay receives this value
 * unchanged.
 */
export interface ConflictResolution {
  readonly outcome: ConflictResolutionOutcome;
  readonly conflictId: string;
  readonly resolutionEventId: string;
  readonly resolutionKind: ConflictResolutionKind;
  readonly resultingVersionId: string | null;
  readonly successorConflictId: string | null;
  readonly completedAt: string;
}

/**
 * One strict explicit-resolution request: the new resolution event
 * identity with its fresh canonical idempotency key, the closed choice,
 * the reviewed remote version and — only under `save_merged` — the
 * verified object reference of an already-uploaded merged result. Never
 * raw bytes, a digest, a locator or any content.
 */
export interface ConflictResolveInput {
  readonly conflictId: string;
  readonly resolutionEventId: string;
  readonly idempotencyKey: string;
  readonly resolutionKind: ConflictResolutionKind;
  readonly reviewedRemoteVersionId?: string | null | undefined;
  readonly verifiedCandidateObjectId?: string | null | undefined;
}

/**
 * The verified read of one immutable evidence role: the exact response
 * bytes after their declared length was verified, plus the declared
 * canonical media type and byte length. The evidence wire surface
 * carries no digest header, so the verification here is the exact
 * content-length; the controller (Task 8) decodes the bytes only for
 * supported text/Markdown media types.
 */
export interface VerifiedConflictEvidence {
  readonly bytes: Uint8Array;
  readonly mediaType: string;
  readonly sizeBytes: number;
}

// --- the closed contract failure -------------------------------------------------------------------

/**
 * The single closed reason of every decoded-contract rejection: an
 * unknown enum value, an ill-typed member, a malformed identity or
 * timestamp, or a foreign payload shape. Raw response detail never
 * reaches the thrown failure.
 */
export type ConflictContractErrorReason = "conflict_contract_invalid";

/** One decoded-contract violation: the closed reason and a static message. */
export class ConflictContractError extends Error {
  readonly reason: ConflictContractErrorReason;

  constructor() {
    super("conflict contract invalid");
    this.name = "ConflictContractError";
    this.reason = "conflict_contract_invalid";
  }
}

function contractInvalid(): ConflictContractError {
  return new ConflictContractError();
}

// --- decoding helpers --------------------------------------------------------------------------------

const UUID_PATTERN = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}

function requireRecord(value: unknown): Record<string, unknown> {
  if (!isRecord(value) || Array.isArray(value)) {
    throw contractInvalid();
  }
  return value;
}

function decodeClosedToken<T extends string>(
  value: unknown,
  closedSet: readonly T[],
): T {
  if (typeof value !== "string" || !(closedSet as readonly string[]).includes(value)) {
    throw contractInvalid();
  }
  // Membership is proven against the closed set above; the cast only
  // refines the narrowed string to its literal union.
  return value as T;
}

function decodeUuid(value: unknown): string {
  if (typeof value !== "string" || !UUID_PATTERN.test(value)) {
    throw contractInvalid();
  }
  return value;
}

function decodeNullableUuid(value: unknown): string | null {
  if (value === null) {
    return null;
  }
  return decodeUuid(value);
}

function decodeTimestamp(value: unknown): string {
  if (typeof value !== "string" || !Number.isFinite(Date.parse(value))) {
    throw contractInvalid();
  }
  return value;
}

function decodeNullableTimestamp(value: unknown): string | null {
  if (value === null) {
    return null;
  }
  return decodeTimestamp(value);
}

function decodeResolutionKind(value: unknown): ConflictResolutionKind {
  return decodeClosedToken(value, CONFLICT_RESOLUTION_KINDS);
}

// --- the record decoders -------------------------------------------------------------------------------

/** Decode one wire `SourceConflictData` onto the frozen plugin shape. */
export function decodeConflictSummary(value: unknown): ConflictSummary {
  const data = requireRecord(value);
  const summary: ConflictSummary = {
    conflictId: decodeUuid(data["conflict_id"]),
    sourceId: decodeNullableUuid(data["source_id"]),
    conflictKind: decodeClosedToken(data["conflict_kind"], CONFLICT_KINDS),
    status: decodeClosedToken(data["status"], CONFLICT_STATUSES),
    originatingEventId: decodeUuid(data["originating_event_id"]),
    originatingDeviceId: decodeUuid(data["originating_device_id"]),
    baseVersionId: decodeNullableUuid(data["base_version_id"]),
    observedRemoteVersionId: decodeNullableUuid(data["observed_remote_version_id"]),
    candidateKind: decodeClosedToken(data["candidate_kind"], CONFLICT_CANDIDATE_KINDS),
    verifiedCandidateObjectId: decodeNullableUuid(data["verified_candidate_object_id"]),
    capturedAt: decodeTimestamp(data["captured_at"]),
    resolutionKind:
      data["resolution_kind"] === null ? null : decodeResolutionKind(data["resolution_kind"]),
    resolutionEventId: decodeNullableUuid(data["resolution_event_id"]),
    resultingVersionId: decodeNullableUuid(data["resulting_version_id"]),
    successorConflictId: decodeNullableUuid(data["successor_conflict_id"]),
    closedAt: decodeNullableTimestamp(data["closed_at"]),
  };
  return summary;
}

/** Decode one wire `SourceConflictDetailData` with its offered choices. */
export function decodeConflictDetail(value: unknown): ConflictDetail {
  const data = requireRecord(value);
  const choicesWire = data["choices"];
  if (!Array.isArray(choicesWire)) {
    throw contractInvalid();
  }
  const choices = choicesWire.map((choice) => decodeResolutionKind(choice));
  return { ...decodeConflictSummary(data), choices };
}

/** Decode one wire `SourceConflictPageData` with its continuation cursor. */
export function decodeConflictPage(value: unknown): ConflictPage {
  const data = requireRecord(value);
  const conflictsWire = data["conflicts"];
  if (!Array.isArray(conflictsWire)) {
    throw contractInvalid();
  }
  const hasMore = data["has_more"];
  if (typeof hasMore !== "boolean") {
    throw contractInvalid();
  }
  return {
    conflicts: conflictsWire.map((conflict) => decodeConflictSummary(conflict)),
    hasMore,
    nextExclusiveStartConflictId: decodeNullableUuid(data["next_exclusive_start_conflict_id"]),
  };
}

/** Decode one wire `SourceConflictResolutionData` outcome. */
export function decodeConflictResolution(value: unknown): ConflictResolution {
  const data = requireRecord(value);
  return {
    outcome: decodeClosedToken(data["outcome"], CONFLICT_RESOLUTION_OUTCOMES),
    conflictId: decodeUuid(data["conflict_id"]),
    resolutionEventId: decodeUuid(data["resolution_event_id"]),
    resolutionKind: decodeResolutionKind(data["resolution_kind"]),
    resultingVersionId: decodeNullableUuid(data["resulting_version_id"]),
    successorConflictId: decodeNullableUuid(data["successor_conflict_id"]),
    completedAt: decodeTimestamp(data["completed_at"]),
  };
}

// --- the resolve input grammar --------------------------------------------------------------------------

/**
 * The canonical lowercase hyphenated UUID text the server's idempotency
 * key grammar accepts — exactly the form the plugin mints with
 * `crypto.randomUUID`.
 */
const IDEMPOTENCY_KEY_PATTERN = UUID_PATTERN;

/**
 * Validate one explicit-resolution request against the server's own field
 * grammar BEFORE any transport contact: canonical UUID identities, the
 * canonical idempotency key, a closed choice, and the verified-object
 * shape per kind (`save_merged` requires the reference; every other kind
 * must not carry one). A violation throws the closed contract error; the
 * request never leaves the device.
 */
export function validateConflictResolveInput(input: ConflictResolveInput): void {
  const resolutionKind = decodeClosedToken(input.resolutionKind, CONFLICT_RESOLUTION_KINDS);
  const reviewedRemoteVersionId = input.reviewedRemoteVersionId ?? null;
  const verifiedCandidateObjectId = input.verifiedCandidateObjectId ?? null;
  if (
    !UUID_PATTERN.test(input.conflictId) ||
    !UUID_PATTERN.test(input.resolutionEventId) ||
    typeof input.idempotencyKey !== "string" ||
    !IDEMPOTENCY_KEY_PATTERN.test(input.idempotencyKey) ||
    (reviewedRemoteVersionId !== null && !UUID_PATTERN.test(reviewedRemoteVersionId)) ||
    (verifiedCandidateObjectId !== null && !UUID_PATTERN.test(verifiedCandidateObjectId))
  ) {
    throw contractInvalid();
  }
  if (resolutionKind === "save_merged") {
    if (verifiedCandidateObjectId === null) {
      throw contractInvalid();
    }
    return;
  }
  if (verifiedCandidateObjectId !== null) {
    throw contractInvalid();
  }
}

// --- the durable no-byte repair vocabularies (journal schema v9) ------------------------------------------

/**
 * The closed target-action vocabulary of one parked local apply: WHICH
 * canonical outcome the Vault still owes. The repair worker re-reads the
 * conflict detail over the wire for the winner identity, so no version
 * id, path or locator is durable here.
 */
export const CONFLICT_LOCAL_REPAIR_ACTIONS = [
  "apply_remote_version",
  "apply_resulting_version",
  "apply_remote_tombstone",
] as const;

/** One parked local apply's closed target action. */
export type ConflictLocalRepairAction = (typeof CONFLICT_LOCAL_REPAIR_ACTIONS)[number];

/**
 * The closed safe-reason vocabulary of one parked local apply: where the
 * local apply stalled. `resolution_committed` parks the fact the moment
 * the canonical resolution committed (the pre-apply crash window);
 * `winner_download_failed` and `vault_apply_failed` name the two bounded
 * retry stages of the local apply itself.
 */
export const CONFLICT_LOCAL_REPAIR_SAFE_REASONS = [
  "resolution_committed",
  "winner_download_failed",
  "vault_apply_failed",
] as const;

/** One parked local apply's closed safe reason. */
export type ConflictLocalRepairSafeReason = (typeof CONFLICT_LOCAL_REPAIR_SAFE_REASONS)[number];

/**
 * One durable pending-local-apply fact of journal schema v9: exactly the
 * conflict UUID, the resolution event identity, the closed target action,
 * the closed safe reason and the retry bookkeeping. The member set makes
 * byte, path and locator storage unrepresentable — evidence and merge
 * drafts live only in bounded ephemeral memory (Task 8).
 */
export interface PendingLocalApply {
  readonly conflictId: string;
  readonly resolutionEventId: string;
  readonly targetAction: ConflictLocalRepairAction;
  readonly safeReason: ConflictLocalRepairSafeReason;
  readonly attemptCount: number;
  readonly nextEligibleRetryEpochMs: number | null;
  readonly createdAtEpochMs: number;
  readonly updatedAtEpochMs: number;
}
